File: Ch01/count_change.erl
-module(count_change).
-author("Kevin C. Baird").
-purpose("SICP in Erlang: Count the number of ways to make change").
-export([count_change/1]).
-include_lib("eunit/include/eunit.hrl").

% Assume US coinage (five types of coins), and call hidden cc() function.
count_change(Amount) -> cc(Amount, 5).

cc(0, _) -> 1; % There is only one (empty) way to make change for no money at all.
cc(_, 0) -> 0; % There is no way to make change when there are no coin types available.

cc(Amount, _) when (Amount < 0) -> 0; % There is no way to make change for a negative amount of money.

% In all other cases...
cc(Amount, Kinds_Of_Coins) ->
  cc(Amount, Kinds_Of_Coins - 1) +
  cc(Amount - first_denomination(Kinds_Of_Coins), Kinds_Of_Coins).

% Define the coinage set as a hidden function.
first_denomination(1) -> 1;
first_denomination(2) -> 5;
first_denomination(3) -> 10;
first_denomination(4) -> 25;
first_denomination(5) -> 50.

%% TESTS

basic_result_test() -> ?assertEqual(292, count_change(100)).

File: Ch01/count_change.exs
# SICP in Elixir: Count the number of ways to make change

defmodule Change do
  def count(amount), do: count(amount, 5)

  # Private functions

  # There is only one (empty) way to make change for no money at all.
  defp count(0, _), do: 1
  # There is no way to make change when there are no coin types available.
  defp count(_, 0), do: 0
  # There is no way to make change for a negative amount of money.
  defp count(amount, _) when amount < 0, do: 0
  # in all other cases...
  defp count(amount, kinds_of_coins) do
    count(amount, kinds_of_coins - 1) +
      count(amount - first_denomination(kinds_of_coins), kinds_of_coins)
  end

  # Assume US coinage
  defp first_denomination(1), do: 1
  defp first_denomination(2), do: 5
  defp first_denomination(3), do: 10
  defp first_denomination(4), do: 25
  defp first_denomination(5), do: 50
end

IO.puts(Change.count(100))
